contracts/interchain-token-service/src/contract/execute.rs

```diff
@@ -186,16 +209,179 @@ pub fn set_chain_config(
+        storage,
+        message_direction.into(),
+        message.token_id(),
+        &token_chain_info,
+    )
+    .change_context(Error::State)
+}
+
+/// Ensures that the token is being redeployed from the same origin chain.
+/// If it's being deployed for the first time, the from chain is saved as the origin chain in the token config.
+fn token_deployment_origin_chain_check(
+    storage: &mut dyn Storage,
+    message: &Message,
+    message_direction: &MessageDirection,
+    token_chain_info: &Option<TokenChainInfo>,
+) -> Result<(), Error> {
+    // Token cannot be redeployed to the destination chain
+    if !matches!(
+        message,
+        Message::DeployInterchainToken { .. } | Message::DeployTokenManager { .. }
+    ) || !matches!(message_direction, MessageDirection::From(_))
+    {
+        return Ok(());
+    }
+
+    let token_config =
+        state::may_load_token_config(storage, &message.token_id()).change_context(Error::State)?;
+
+    if let Some(TokenConfig { origin_chain, .. }) = token_config {
+        // Token can only be redeployed from the same origin chain
+        ensure!(
+            origin_chain == ChainNameRaw::from(message_direction.clone()),
+            Error::TokenDeployedFromNonOriginChain {
+                token_id: message.token_id(),
+                origin_chain,
+                chain: message_direction.clone().into(),
+            }
+        );
+
+        // Token chain info must already exist from previous deployment
+        ensure!(
+            token_chain_info.is_some(),
+            Error::TokenNotDeployed {
+                token_id: message.token_id(),
+                chain: message_direction.clone().into(),
+            }
+        );
+    } else {
+        // Token is being deployed for the first time
+        let token_config = TokenConfig {
+            origin_chain: ChainNameRaw::from(message_direction.clone()),
+        };
+
+        state::save_token_config(storage, &message.token_id(), &token_config)
+            .change_context(Error::State)?;
+
+        ensure!(
+            token_chain_info.is_none(),
+            Error::TokenAlreadyDeployed {
+                token_id: message.token_id(),
+                chain: message_direction.clone().into(),
+            }
+        );
+    }
+
+    Ok(())
+}
+
+/// Ensures that the token is not being redeployed to the same destination chain.
+fn token_redeployment_check(
+    message: &Message,
+    message_direction: &MessageDirection,
+    token_chain_info: &Option<TokenChainInfo>,
+) -> Result<(), Error> {
+    if matches!(
+        message,
+        Message::DeployInterchainToken { .. } | Message::DeployTokenManager { .. }
+    ) && matches!(message_direction, MessageDirection::To(_))
+        && token_chain_info.is_some()
+    {
+        bail!(Error::TokenAlreadyDeployed {
+            token_id: message.token_id(),
+            chain: message_direction.clone().into(),
+        });
+    }
+
+    Ok(())
```

Comment:
```suggestion
    match (message, message_direction, token_chain_info) {
        (
            Message::DeployInterchainToken { .. } | Message::DeployTokenManager { .. },
            MessageDirection::To(_),
            Some(_),
        ) => {
            bail!(Error::TokenAlreadyDeployed {
                token_id: message.token_id(),
                chain: message_direction.clone().into(),
            });
        }
        _ => Ok(()),
    }
```
Chain of `&&` is very hard to read.

Reply:
I didn't find either to be too clean, but i can go with whatever folks prefer